src/latbin/lattice-compose-fsts.cc

```diff
@@ -0,0 +1,194 @@
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "fstext/fstext-lib.h"
+#include "lat/kaldi-lattice.h"
+#include "lat/lattice-functions.h"
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    typedef kaldi::int32 int32;
+    typedef kaldi::int64 int64;
+    using fst::SymbolTable;
+    using fst::VectorFst;
+    using fst::StdArc;
+
+    const char *usage =
+        "Composes lattices (in transducer form, as type Lattice) with word-network FSTs.\n"
+        "Either with a single FST from rxfilename or with per-utterance FSTs from rspecifier.\n"
+        "The FST weights are interpreted as \"graph weights\" when converted into the Lattice format.\n"
+        "\n"
+        "Usage: lattice-compose-fsts [options] lattice-rspecifier1 "
+        "(fst-rspecifier2|fst-rxfilename2) lattice-wspecifier\n"
+        " e.g.: lattice-compose-fsts ark:1.lats ark:2.fsts ark:composed.lats\n"
+        " or: lattice-compose-fsts ark:1.lats G.fst ark:composed.lats\n";
+
+    ParseOptions po(usage);
+
+    bool write_compact = true;
+    int32 num_states_cache = 50000;
+    int32 phi_label = fst::kNoLabel; // == -1
+    po.Register("write-compact", &write_compact, "If true, write in normal (compact) form.");
+    po.Register("phi-label", &phi_label, "If >0, the label on backoff arcs of the LM");
+    po.Register("num-states-cache", &num_states_cache,
+                "Number of states we cache when mapping LM FST to lattice type. "
+                "More -> more memory but faster.");
+    po.Read(argc, argv);
+
+    if (po.NumArgs() != 3) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+    KALDI_ASSERT(phi_label > 0 || phi_label == fst::kNoLabel); // e.g. 0 not allowed.
+
+    std::string lats_rspecifier1 = po.GetArg(1),
+        arg2 = po.GetArg(2),
+        lats_wspecifier = po.GetArg(3);
+    int32 n_done = 0, n_fail = 0;
+
+    SequentialLatticeReader lattice_reader1(lats_rspecifier1);
+
+    CompactLatticeWriter compact_lattice_writer;
+    LatticeWriter lattice_writer;
+
+    if (write_compact)
+      compact_lattice_writer.Open(lats_wspecifier);
+    else
+      lattice_writer.Open(lats_wspecifier);
+
+    if (ClassifyRspecifier(arg2, NULL, NULL) == kNoRspecifier) {
+      std::string fst_rxfilename = arg2;
+      VectorFst<StdArc>* fst2 = fst::ReadFstKaldi(fst_rxfilename);
+      // mapped_fst2 is fst2 interpreted using the LatticeWeight semiring,
+      // with all the cost on the first member of the pair (since we're
+      // assuming it's a graph weight).
+      if (fst2->Properties(fst::kILabelSorted, true) == 0) {
+        // Make sure fst2 is sorted on ilabel.
+        fst::ILabelCompare<StdArc> ilabel_comp;
+        ArcSort(fst2, ilabel_comp);
+      }
+      /* // THIS MAKES ALL STATES FINAL STATES! WHY?
```

Comment:
Fishy line

Reply:
Does the ρ-matcher treats ε specially? I vaguely remember it doesn't. If this is so, and it falls under the "everything else" category, I can guess why it might happen.

Reply:
@galv What is wrong with that ? Do you react on the present `ArcSort` or the disabled `PropagateFinal` ?

Reply:
He may have been referring to the comment 
// THIS MAKES ALL STATES FINAL STATES! WHY?
... 
It is only relevant if the --phi-label option was specified (0 is not allowed for phi label),
 in which case it will use backoff semantics.  
I think it makes sense to keep that commented-out code since the phi stuff is supported.
It is intended that all states in G.fst should be final, because EOS is allowed from all states.

Reply:
.. PropagateFinal() fixes the semantics of EOS, because it is done as a final-prob, not an arc, so the phi-matcher does not naturally handle it.   (In k2 this kind of problem could not happen because final-probs are done as a special transition to the superfinal-state, so we wouldn't need special-purpose code).

Reply:
Hi @vesis84 the main concern is that you are trying to commit code with a commented out block that you apparently don't understand, from a software engineering point of view.

Also, I thought it was odd that you say that PropagateFinal makes all states final, but fst2 is expected to be a single-state FST based on what you wrote. In this case, I would expect you to set the only state in the FST to be final anyway. Unfortunately, I don't know how phi-symbol matching works, though, so I can't really say more about what your specific problem may be.

Reply:
In FST composition with ρ-matcher the ε symbols the output of the composition should be the same regardless if ε is removed or not. On the other hand, for me the runtime of composition was faster, when i removed ε from the boosting graph B in the HCLG o B composition.